File: ios/TCPClientModule.swift
//
//  LocalMessagingClient.swift
//  LocalServer
//
//  Created by David Lang on 02.06.2022.
//  Copyright © 2022 Figure, Inc. All rights reserved.
//
import Foundation

@objc(TCPClientModule)
class TCPClientModule: RCTEventEmitter {

    private let eventNames: [String]! = TCPClientEventName.allValues
    private let eventEmitter: EventEmitterWrapper = EventEmitterWrapper()
    
    private var clients: [String: TCPClient] = [:]
    
    override init() {
        super.init()
        eventEmitter.setEventEmitter(eventEmitter: self)
    }
    
    @objc(createClient:withHost:withPort:withResolver:withRejecter:)
    func createClient(id: String, host: String, port: UInt16, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("TCPClientModule - createClient - started")
        if (clients[id] != nil) {
            reject("client.already-exists", "Client with this id already exists", nil)
            return
        }
        let client: TCPClient = TCPClient(id: id, host: host, port: port, eventEmitter: eventEmitter)
        clients[id] = client
        client.start()
        resolve(true)
    }

    @objc(stopClient:withResolver:withRejecter:)
    func stopClient(id: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("TCPClientModule - stopClient - started")
        if let client: TCPClient = clients[id] {
            client.stop()
            clients.removeValue(forKey: id)
            resolve(true)
        } else {
            reject("client.not-exists", "Client with this id does not exist", nil)
        }
    }

    @objc(send:withMessage:withResolver:withRejecter:)
    func send(clientId: String, message: String, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        print("TCPClientModule - send - started")
        if let client: TCPClient = clients[clientId] {
            client.send(message: message)
            resolve(true)
        } else {
            reject("client.not-exists", "Client with this id does not exist", nil)
        }
    }

    override func supportedEvents() -> [String]! {
        return self.eventNames
    }
}

File: android/src/main/java/com/reactnativelocalserver/TCPClientModule.java
package com.reactnativelocalserver;

import android.util.Log;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;
import com.reactnativelocalserver.tcp.TCPClient;
import com.reactnativelocalserver.tcp.factory.TCPClientFactory;
import com.reactnativelocalserver.utils.EventEmitter;
import com.reactnativelocalserver.utils.StopReasonEnum;

import java.net.UnknownHostException;
import java.util.HashMap;
import java.util.Map;

@ReactModule(name = TCPClientModule.NAME)
public class TCPClientModule extends ReactContextBaseJavaModule {
    public static final String NAME = "TCPClientModule";

    private final TCPClientFactory clientFactory;

    private final EventEmitter eventEmitter;

    private final Map<String, TCPClient> clients = new HashMap();

    public TCPClientModule(ReactApplicationContext reactContext, EventEmitter eventEmitter, TCPClientFactory clientFactory) {
        super(reactContext);
        this.eventEmitter = eventEmitter;
        this.clientFactory = clientFactory;
    }

    public Map<String, TCPClient> getClients() {
        return clients;
    }

    @Override
    @NonNull
    public String getName() {
        return NAME;
    }

    @ReactMethod
    public void createClient(String id, String host, int port, Promise promise) {
        Log.d(NAME, "createClient started for id:" + id);
        if (clients.get(id) != null) {
            promise.reject("tcp.client.already-exists", "Client with this id already exists");
            return;
        }
        try {
            TCPClient client = clientFactory.of(id, host, port, eventEmitter);
            client.setOnConnectionClosed(this::onConnectionClosed);
            client.start();
            clients.put(id, client);
            promise.resolve(true);
        } catch (UnknownHostException e) {
            promise.reject("tcp.client.error", "unknown host: " + host);
        } catch (Exception e) {
            promise.reject("tcp.client.error", e.getMessage());
        }
    }

    @ReactMethod
    public void stopClient(String id, String reason, Promise promise) {
        Log.d(NAME, "stopClient started for id:" + id);
        TCPClient client = clients.get(id);
        if (client == null) {
            promise.reject("tcp.client.not-exists", "Client with this id does not exist");
            return;
        }
        try {
            client.stop(reason);
            clients.remove(id);
            promise.resolve(true);
        } catch (Exception e) {
            promise.reject("tcp.client.error", e.getMessage());
        }
    }

    @ReactMethod
    public void send(String clientId, String message, Promise promise) {
        Log.d(NAME, "send started for client:" + clientId);
        TCPClient client = clients.get(clientId);
        if (client == null) {
            promise.reject("tcp.client.not-exists", "Client with this id does not exist");
            return;
        }
        try {
            client.send(message);
            promise.resolve(true);
        } catch (Exception e) {
            promise.reject("tcp.client.error", e.getMessage());
        }
    }

    @Override
    public void invalidate() {
        Log.d(NAME, "invalidate - number of clients: " + clients.size());
        for (Map.Entry<String, TCPClient> entry : clients.entrySet()) {
            try {
                entry.getValue().stop(StopReasonEnum.Invalidation);
            } catch (Exception e) {
                Log.e(NAME, "invalidate - failed to stop client: " + entry.getKey(), e);
            }
        }
        clients.clear();
        super.invalidate();
    }

    private void onConnectionClosed(String clientId) {
        clients.remove(clientId);
    }
}
